var/spack/repos/builtin/packages/gdal/package.py

```diff
@@ -237,370 +405,282 @@ def patch(self):
-            # https://trac.osgeo.org/gdal/wiki/Epsilon
-            args.append('--with-epsilon=no')
-
-        # TODO: add packages for these dependencies (only for 3.1 and older)
-        if spec.satisfies('@:3.1'):
-            # https://trac.osgeo.org/gdal/wiki/ArcSDE
-            args.append('--with-sde=no')
-
-        # TODO: add packages for these dependencies (only for 2.3 and older)
-        if spec.satisfies('@:2.3'):
-            args.append('--with-php=no')
-
-        # TODO: add packages for these dependencies (only for 3.2 and newer)
-        if spec.satisfies('@3.2:'):
-            args.append('--with-heif=no')
-
-        # TODO: add packages for these dependencies (only for 3.1 and newer)
-        if spec.satisfies('@3.1:'):
             args.extend([
-                '--with-exr=no',
-                '--with-rdb=no',
+                self.with_or_without(
+                    'xml2', variant='libxml2', package='libxlm2', attribute='command'
+                ),
+                self.with_or_without('mongocxx', variant='mongocxx'),
+                self.with_or_without(
+                    'pg', variant='postgresql',
+                    package='postgresql', attribute='command'
+                ),
             ])
 
-        # TODO: add packages for these dependencies (only for 3.0 and newer)
-        if spec.satisfies('@3.0:'):
-            args.extend([
-                '--with-tiledb=no',
-                '--with-mongocxxv3=no',
-            ])
+        if self.spec.satisfies('@2.3:'):
+            args.append('--with-proj={}'.format(self.spec['proj'].prefix))
 
-        # TODO: add packages for these dependencies (only for 2.3 and newer)
-        if spec.satisfies('@2.3:'):
-            args.extend([
-                '--with-jp2lura=no',
-                '--with-rasterlite2=no',
-                # https://trac.osgeo.org/gdal/wiki/DxfDwg
-                '--with-teigha=no',
-                '--with-sfcgal=no',
-            ])
+        if self.spec.satisfies('@2.2:'):
+            args.append(self.with_or_without('teigha', package='teigha'))
+        else:
+            args.append(
+                self.with_or_without('dwgdirect', variant='teigha', package='teigha')
+            )
 
-        # TODO: add packages for these dependencies (only for 2.1 and newer)
-        if spec.satisfies('@2.1:'):
-            args.extend([
-                '--with-mongocxx=no',
-                '--with-pdfium=no',
-            ])
+        if '+hdf4' in self.spec:
+            hdf4 = self.spec['hdf']
+            if '+external-xdr' in hdf4 and hdf4['rpc'].name != 'libc':
+                args.append('LIBS=' + hdf4['rpc'].libs.link_flags)
 
-        if libs:
-            args.append('LIBS=' + ' '.join(libs))
+        # Remove empty strings
+        args = [arg for arg in args if arg]
 
         return args
 
-    # https://trac.osgeo.org/gdal/wiki/GdalOgrInJavaBuildInstructionsUnix
+    @when('@:3.4')
+    def cmake(self, spec, prefix):
```

Comment:
IMO it's worth adding a comment here that we are effectively overriding the `cmake` phase to do non-CMake things for older versions.

Possibly clearer (though not necessary to do here): if `phases` is implemented as a property which omits `cmake` for `@:3.4`, do you think that would be easier to work with?

(as a side note, we have ongoing discussions for multi-build-system packages, so this should also become easier then, and is another reason it may not be worth improving the situation now).

Reply:
@alalazo what are your thoughts on this? We have hundreds of other packages that override phases for historical build systems without comments explaining them. I'm inclined to wait until multi-build system packages have better support.

Reply:
I think we can consider this good. I brought it up because it took me a minute to figure out what was going on. I think in the future (for other cases where you have to do this) it would be useful to have the comment (but I agree that a systematic insertion of this comment would not be productive, in particular because we aim to make this type of thing unnecessary).